src/main/scala/firrtl/passes/ConstraintSolver.scala

```diff
@@ -0,0 +1,267 @@
+    val checkMap = new mutable.HashMap[String, Constraint]()
+    constraints.foldLeft(Seq[Constraint]()) { (seq, c) =>
+      checkMap.get(c.left) match {
+        case None =>
+          checkMap(c.left) = c
+          seq ++ Nil
+        case Some(x) if x.geq != c.geq => seq ++ Seq(x, c)
+        case Some(x) => seq ++ Nil
+      }
+    }
+  }
+  def getDepth(i: IsConstrainable): Int = i.children match {
+    case Nil => 1
+    case other => other.map(getDepth).max + 1
+  }
+  def getSize(i: IsConstrainable): Int = i.children match {
+    case Nil => 1
+    case other => other.map(getSize).sum
+  }
+
+  def solve() = {
+    // Check that all constraints per name are either geq or leq, but not both
+    val illegals = check()
+    if (illegals != Nil) sys.error("Illegal constraints!: $illegals")
+
+    // Combines constraints on the same VarWidth into the same constraint
+    val uniqueConstraints = mergeConstraints(constraints.toSeq)
+    //println("Unique Constraints!")
+    //println(uniqueConstraints.mkString("\n"))
+
+    /* Forward solve
+     * Returns a solved list where each constraint undergoes:
+     *  1) Continuous Solving (using triangular solving)
+     *  2) Remove Cycles
+     *  3) Move to solved if not self-recursive
+     */
+    val forwardConstraintMap = new ConstraintMap
+    val orderedVars = mutable.HashMap[Int, String]()
+
+    /* Keep track of size of expression to time to optimize */
+    val subMap = mutable.HashMap[Int, (Int, Double)]()
+    val optMap = mutable.HashMap[Int, (Int, Double)]()
+    val rmcMap = mutable.HashMap[Int, (Int, Double)]()
+    def update(map: mutable.HashMap[Int, (Int, Double)], time: Double, size: Int): Unit = {
+      val (n, prev) = map.getOrElse(size, (0, 0.0))
+      map(size) = (n + 1, prev + time)
+    }
+
+    //val (total, result) = time {
+      var index = 0
+      for (constraint <- uniqueConstraints) {
+        //TODO: Risky if used improperly... need to check whether substitution from a leq to a geq is negated (always).
+
+        //val (timesub, subbedRight) = time { substitute(forwardConstraintMap)(constraint.right) }
+        val subbedRight = substitute(forwardConstraintMap)(constraint.right)
+        //update(subMap, timesub, getDepth(subbedRight))
+
+
+        //val (timeopt, optSubbedRight) = time { subbedRight.optimize() }
+        val optSubbedRight = subbedRight//.optimize()
+        //update(optMap, timeopt, getDepth(subbedRight))
+
+        val name = constraint.left
+        //val (timermc, finishedRight) = time { removeCycle(name, constraint.geq)(subbedRight)}
+        val finishedRight = removeCycle(name, constraint.geq)(subbedRight)
+        //update(rmcMap, timermc, getDepth(finishedRight))
+        if (!hasVar(name)(finishedRight)) {
+          forwardConstraintMap(name) = (finishedRight, constraint.geq)
+          orderedVars(index) = name
+          index += 1
+        }
+      }
+    //}
+    //println(s"All Together (ms): $total")
```

Comment:
Should these commented out `println`s be logged?

Reply:
Could we turn it into a flag that's disabled by default? That way the code wiould be a bit more immune to code rot if the surrounding variables change, compared to if it's just a comment.

Reply:
Meh I just got rid of them. Its better to debug via flamegraphs and breakpoints anyways.